go-controller/pkg/cni/helper_linux.go

```diff
@@ -54,48 +94,112 @@ func setupInterface(netns ns.NetNS, containerID, ifName, macAddress, ipAddress,
+	}
+
+	return hostIface, contIface, nil
+}
+
+// Setup sriov interface in the pod
+func setupSriovInterface(netns ns.NetNS, containerID, ifName, macAddress, ipAddress, gatewayIP string, mtu int, pciAddrs string) (*current.Interface, *current.Interface, error) {
+	hostIface := &current.Interface{}
+	contIface := &current.Interface{}
+
+	// 1. get VF netdevice from PCI
+	vfNetdevices, err := sriovnet.GetNetDevicesFromPci(pciAddrs)
+	if err != nil {
+		return nil, nil, err
+	}
+	// Make sure we have 1 netdevice per pci address
+	if len(vfNetdevices) != 1 {
+		return nil, nil, fmt.Errorf("failed to get one netdevice interface per %s", pciAddrs)
+	}
+	vfNetdevice := vfNetdevices[0]
+	// 2. get Uplink netdevice
+	uplink, err := sriovnet.GetUplinkRepresentor(pciAddrs)
+	if err != nil {
+		return nil, nil, err
+	}
+	// 3. get VF index from PCI
+	vfIndex, err := sriovnet.GetVfIndexByPciAddress(pciAddrs)
+	if err != nil {
+		return nil, nil, err
+	}
+	// 4. lookup representor
+	rep, err := sriovnet.GetVfRepresentor(uplink, vfIndex)
+	if err != nil {
+		return nil, nil, err
+	}
+	oldHostRepName := rep
+	// rename the host Vf representor
+	hostIface.Name = containerID[:15]
+	if err = renameLink(oldHostRepName, hostIface.Name); err != nil {
+		return nil, nil, fmt.Errorf("failed to rename %s to %s: %v", oldHostRepName, hostIface.Name, err)
+	}
+	link, err := netlink.LinkByName(hostIface.Name)
+	if err != nil {
+		return nil, nil, err
+	}
+	hostIface.Mac = link.Attrs().HardwareAddr.String()
+
+	// 6. Move VF to Container namespace
+	err = moveIfToNetns(vfNetdevice, netns)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = netns.Do(func(hostNS ns.NetNS) error {
+		contIface.Name = ifName
+		err = renameLink(vfNetdevice, contIface.Name)
 		if err != nil {
 			return err
 		}
-		err = netlink.AddrAdd(link, addr)
+		link, err = netlink.LinkByName(contIface.Name)
 		if err != nil {
-			return fmt.Errorf("failed to add IP addr %s to %s: %v", ipAddress, contIface.Name, err)
+			return err
 		}
-
-		gw := net.ParseIP(gatewayIP)
-		if gw == nil {
-			return fmt.Errorf("parse ip of gateway failed")
+		err = netlink.LinkSetUp(link)
+		if err != nil {
+			return err
 		}
-		err = ip.AddRoute(nil, gw, link)
+
+		err = setupNetwork(link, macAddress, ipAddress, gatewayIP)
 		if err != nil {
 			return err
 		}
 
-		oldHostVethName = hostVeth.Name
+		contIface.Mac = macAddress
+		contIface.Sandbox = netns.Path()
 
 		return nil
 	})
 	if err != nil {
 		return nil, nil, err
 	}
 
-	// rename the host end of veth pair
-	hostIface.Name = containerID[:15]
-	if err := renameLink(oldHostVethName, hostIface.Name); err != nil {
-		return nil, nil, fmt.Errorf("failed to rename %s to %s: %v", oldHostVethName, hostIface.Name, err)
+	// rename the VF representor
+	if err := renameLink(hostIface.Name, containerID[:15]); err != nil {
+		return nil, nil, fmt.Errorf("failed to rename %s to %s: %v", hostIface.Name, containerID[:15], err)
 	}
+	hostIface.Name = containerID[:15]
 
```

Comment:
L198-202 is already done at L152-156. So, please remove the duplicate code.

Reply:
yes you are right, I will remove it.